website/docs/language/expressions/custom-conditions.mdx

```diff
@@ -0,0 +1,353 @@
+
+  # The EC2 instance will have an encrypted root volume.
+  precondition {
+    condition     = data.aws_ebs_volume.example.encrypted
+    error_message = "The server's root volume is not encrypted."
+  }
+}
+```
+
+### Choosing Between Preconditions and PostConditions
+
+You can often implement a validation check as either a postcondition of the resource producing the data or as a precondition of a resource or output value using the data. To decide which is most appropriate, consider whether the check is representing either an assumption or a guarantee.
+
+| Type | Definition | Expression to Use |
+| -----|------------|--------------|
+| Assumption | A condition that must be true in order for the configuration of a particular resource to be usable. For example, an `aws_instance` configuration can have the assumption that the given AMI will always be configured for the `x86_64` CPU architecture. | Preconditions, so that future maintainers can find them close to the other expressions that rely on that condition. This lets them understand more about what that resource is intended to allow.|
+| Guarantee | A characteristic or behavior of an object that the rest of the configuration should be able to rely on. For example, an `aws_instance` configuration can have the guarantee that an EC2 instance will be running in a network that assigns it a private DNS record. | Postconditions, so that future maintainers can find them close to the resource configuration that is responsible for implementing those guarantees. This lets them more easily determine which behaviors they should preserve when changing the configuration. |
+
+We recommend also considering the following factors.
+
+- Which resource or output value would be most helpful to report in the error message. Terraform will always report errors in the location where the condition was declared.
+- Which approach is more convenient. If a particular resource has many dependencies that all make an assumption about that resource, it can be pragmatic to declare that once as a post-condition of the resource, rather than declaring it many times as preconditions on each of the dependencies.
+- Whether it is helpful to declare the same or similar conditions as both preconditions and postconditions. This can be useful if the postcondition is in a different module than the precondition because it lets the modules verify one another as they evolve independently.
+
+
+## Condition Expressions
+
+Input variable validation, preconditions, and postconditions all require a `condition` argument. This is a boolean expression that should return `true` if the intended assumption or guarantee is fulfilled or `false` if it does not.
+
+You can use any of Terraform's built-in functions or language operators
+in a condition as long as the expression is valid and returns a boolean result. The following language features are particularly
+useful when writing condition expressions.
+
+### `contains` Function
```

Comment:
I like this extended section describing various relevant language features.

As a general note it might be helpful to find for each of these a suitable place to link to for more information. For the functions it's relatively clear because each one has its own full page anyway. The operators are trickier because we just describe them all together on a single page [Operators](https://www.terraform.io/language/expressions/operators), but at least for the section "Boolean Operators" we could link across to [Logical Operators](https://www.terraform.io/language/expressions/operators#logical-operators).

(This also made me spot a terminology mismatch; maybe we should change "Boolean Operators" here to "Logical Operators", or vice-versa, depending on which one feels more intuitive. My initial instinct is that "boolean operators" probably gets to the point better, but if we _do_ change the Operators page heading we'll need to leave a hidden anchor in there in case there are already links to the existing section name.)

---

Perhaps we should lead here with the [Comparison Operators](https://www.terraform.io/language/expressions/operators#comparison-operators) before we talk about the more specialized situations, because I expect in practice they'll probably be the most common starting point. We already made plenty of use of them in the examples so far on this page, so their absence here stands out.

Reply:
good call - changed "boolean" to logical to match the page, I also moved the logical operators section to be first one and I also directly call out the fact that you can use other types of operators and link to that docs page :-) Thank you for the good suggestions!